### PR TITLE
🐛 Suppress legit OS errors on socket shutdown

### LINKED ISSUE #341: [8.4.8 regression?] ConnectionResetError on Windows
❓ **I'm submitting a ...**
- [X] 🐞 bug report
- [ ] 🐣 feature request
- [ ] ❓ question about the decisions made in the repository

🐞 **Describe the bug. What is the current behavior?**
After updating from 8.4.7 to .8, in around 50% of the testsuite runs of [my project](https://www.qutebrowser.org/) (which runs end to end tests of a webbrowser against cheroot), I get this traceback on Windows:

```python-traceback
Exception in thread CP Server Thread-6:
Traceback (most recent call last):
  File "c:\hostedtoolcache\windows\python\3.7.9\x64\lib\threading.py", line 926, in _bootstrap_inner
    self.run()
  File "D:\a\qutebrowser\qutebrowser\.tox\py37-pyqt515\lib\site-packages\cheroot\workers\threadpool.py", line 130, in run
    conn.close()
  File "D:\a\qutebrowser\qutebrowser\.tox\py37-pyqt515\lib\site-packages\cheroot\server.py", line 1357, in close
    self._close_kernel_socket()
  File "D:\a\qutebrowser\qutebrowser\.tox\py37-pyqt515\lib\site-packages\cheroot\server.py", line 1480, in _close_kernel_socket
    shutdown(socket.SHUT_RDWR)  # actually send a TCP FIN
ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host
```

💡 **To Reproduce**
I'm afraid I don't have a reproducer for this at the moment, and no idea where to start...

📋 **Environment**
- Cheroot version: 8.4.8
- Python version: 3.7.9
- OS: Windows Server 2019 (via GitHub Actions)

📋 **Additional context**
Failing builds:

- https://github.com/qutebrowser/qutebrowser/pull/5924/checks?check_run_id=1473349077
- https://github.com/qutebrowser/qutebrowser/runs/1460847930

I don't remember ever seeing this before 8.4.8, so #337 might be responsible for this somehow? (cc @liamstask)

### PR DESCRIPTION
**Pull request by webknjaz**
When closing a socket we normally want to terminate the transport-
level connection by sending a TCP FIN packet over the wire. This works
great if the client is willing to perform a 3-way handshake but
sometimes the client-side just drops the connection by sending an RST
instead of a FIN. In this case, our server-side `socket.shutdown()`
call will raise an OSError (socket.error on Python 2) or its
derivatives. Which is perfectly fine and it's alright for us to ignore
those.

The kernel socket close helper rewrite introduced a behavior of only
suppressing ENOTCONN in v8.4.8 (the case when the client is no longer
with us) but it left own a few other cases that may be happening too.

This change fixes that by extending the list of errors that are to be
suppressed. Here's what's handled from now on:
  * ENOTCONN — client is no longer connected
  * EPIPE — write on a pipe while the other end has been closed
  * ESHUTDOWN — write on a socket which has been shutdown for writing
  * ECONNRESET — connection is reset by the peer

❓ **What kind of change does this PR introduce?**
  - [x] 🐞 bug fix
  - [ ] 🐣 feature
  - [ ] 📋 docs update
  - [ ] 📋 tests/coverage improvement
  - [ ] 📋 refactoring
  - [ ] 💥 other

📋 **What is the related issue number (starting with `#`)**

Fixes #341

❓ **What is the current behavior?** (You can also link to an open issue here)

Connection drop/reset/shutdown initiated by the client-side while the server attempts to send a TCP FIN triggers an unhandled OSError/socket.error during the socket shutdown and thus leaks into the logs.

❓ **What is the new behavior (if this is a feature change)?**

It's handled gracefully and suppressed.

📋 **Other information**:

Refs:
  * https://en.wikipedia.org/wiki/Transmission_Control_Protocol#Connection_termination
  * http://cs.baylor.edu/~donahoo/practical/CSockets/TCPRST.pdf
  * https://github.com/cherrypy/cheroot/issues/341#issuecomment-735884889

📋 **Checklist**:

  - [x] I think the code is well written
  - [x] I wrote [good commit messages][1]
  - [x] I have [squashed related commits together][2] after the changes have been approved
  - [ ] Unit tests for the changes exist
  - [ ] Integration tests for the changes exist (if applicable)
  - [x] I used the same coding conventions as the rest of the project
  - [x] The new code doesn't generate linter offenses
  - [ ] Documentation reflects the changes
  - [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences

[1]: http://chris.beams.io/posts/git-commit/
[2]: https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

---
This change is [Reviewable](https://reviewable.io/reviews/cherrypy/cheroot/342)

### CODE REVIEW

**Comment by webknjaz** [comment]
cc @liamstask && @The-Compiler

File: cheroot/errors.py [hunk 1]
```diff
@@ -56,3 +57,24 @@ def plat_specific_errors(*errnames):
 if sys.platform == 'darwin':
     socket_errors_to_ignore.extend(plat_specific_errors('EPROTOTYPE'))
     socket_errors_nonblocking.extend(plat_specific_errors('EPROTOTYPE'))
+
+
+acceptable_sock_shutdown_error_codes = {
+    errno.ENOTCONN,
+    errno.EPIPE, errno.ESHUTDOWN,  # corresponds to BrokenPipeError in Python 3
```
Comment (liamstask) [inline cheroot/errors.py:64]: do both of these map to `BrokenPipeError`? maybe worth putting on their own line so the comment is unambiguous

**Review by liamstask — APPROVED** [review]
the change lgtm, not sure of the best way to test

**Review by The-Compiler — APPROVED** [review]
Sorry for the delay! I set up a CI run which runs 10 Windows jobs resulting in 4 of them failing:

- [WIP: test cheroot · qutebrowser/qutebrowser@42c979c](https://github.com/qutebrowser/qutebrowser/actions/runs/396329771)

Then I tested this branch and all jobs pass again:

[test bugfix · qutebrowser/qutebrowser@a6147bf](https://github.com/qutebrowser/qutebrowser/actions/runs/396388021)

So I'm guessing this PR indeed fixes my issue - thanks for the quick fix! :+1:

**Comment by webknjaz** [comment]
Thanks for the feedback! I'll see if I'll be able to come up with some regression tests soonish and if not, I'll just address the suggestions and merge the PR as is.
P.S. I can't promise a quick release though — our automated publishing is affected by #343 and I'm working to fix that.

File: cheroot/errors.py [hunk 2]
```diff
@@ -56,3 +57,24 @@ def plat_specific_errors(*errnames):
...
+* ESHUTDOWN — write on a socket which has been shutdown for writing
+* ECONNRESET — connection is reset by the peer
+
+Ref: https://github.com/cherrypy/cheroot/issues/341#issuecomment-735884889
+"""
+
+try:  # py3
+    broken_connection_exceptions = (BrokenPipeError, ConnectionResetError)
+except NameError:  # py2
+    broken_connection_exceptions = ()
```
[the remark this answers is not in the document]
Comment (webknjaz) [reply cheroot/errors.py:80]: Done!

**Comment by webknjaz** [comment]
This is how one can trigger RST (ConnectionResetError: [Errno 104] Connection reset by peer) https://stackoverflow.com/a/40779258/595220

**Comment by webknjaz** [comment]
I can only reproduce RST on `recv()` calls but not on `shutdown()`, but I was able to consistently cause `OSError(107, 'Transport endpoint is not connected')` in tests... Can it be that `shutdown()` only raises `ECONNRESET` on Windows?

**Comment by webknjaz** [comment]
CPython devs appear to be struggling with reproducers too: https://bugs.python.org/issue30319 / https://bugs.python.org/issue30329 / https://github.com/python/cpython/commit/83a2c28 / https://github.com/python/cpython/blob/c39b52f/Lib/poplib.py#L297-L302 / https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-shutdown.

I guess I'll just try to monkey-patch shutdown to raise states that are hard to reproduce until somebody with Windows can take a look.

**Comment by webknjaz** [comment]
Hey @mxii-ca, I recall you performed an excellent debugging of TLS earlier this year so I thought you may be able to help us out here. Do you know of any ways to trigger TCP RST to reproduce an issue when the client sends RST while the server initiates a shutdown sequence with FIN? Anything Windows-specific maybe?

**Comment by jaraco** [comment]
> I can't promise a quick release though

Releases shouldn't be blocked on automation. I'll cut a manual release if needed.

**Comment by jaraco** [comment]
Since this change fixes an observable regression and as the user was able to confirm this change addresses that regression, I'd like to proceed with merging this fix and getting it out. I've filed #344 to capture the need to follow-up with a proper regression test.

**Comment by webknjaz** [comment]
Oh, I had some local tests that I was about to push today.

**Comment by jaraco** [comment]
Great! Just push them as a separate PR (if you wish), or just push them to maint/8.x, and that'll close out #344.

**Comment by webknjaz** [comment]
I'll work on it. I had an amended commit so I need to deal with that first.

**Comment by webknjaz** [comment]
@jaraco I noticed you had to make a manual release. That should work now that I've applied a hotfix to GHA: https://github.com/cherrypy/cheroot/commit/71e03b1b694dd8bd8e684743dc8881eef3da18aa. Since there's some flaky test jobs, it may fail in the middle of the test stage, though. If that happens, just restarting the workflow should get things going.

**Comment by webknjaz** [comment]
@The-Compiler This is released as v8.5.0 FYI

**Comment by jaraco** [comment]
> I'll work on it. I had an amended commit so I need to deal with that first.

Gah. I hadn't considered that contingency. Let me know if I can help.

**Comment by mxii-ca** [comment]
@webknjaz you can change the behavior of close from the graceful `FIN` shutdown to sending a `RST` by setting the `SO_LINGER` socket option

when `l_onoff` is non-zero (true), the connection will linger open after `close` has been called to allow for send queues to empty.  
however, if the amount of time to wait, `l_linger`, is set to **0**, then instead of waiting, it will hard close the connection by sending a reset

```python
import socket

s = socket.create_connection(("127.0.0.1", 5673))

linger = bytearray()
linger.extend((1).to_bytes(2, "little", signed=False))  # l_onoff
linger.extend((0).to_bytes(2, "little", signed=False))  # l_linger
s.setsockopt(socket.SOL_SOCKET, socket.SO_LINGER, bytes(linger))

# send RST
s.close()
```

this should work cross-platform, not just on Windows

**Comment by mxii-ca** [comment]
unfortunately, using `SO_LINGER` to send a `RST` in the middle of a `FIN` shutdown is nearly impossible as it becomes a timing issue

however, reliably sending a `RST` in the middle of a `FIN` shutdown should be possible with a raw socket; you'd have to write your own TCP stack or modify someone else's to send the `RST` at the desired time.
a quick google search brought up this example for using raw sockets in python:
https://www.binarytides.com/raw-socket-programming-in-python-linux/

note that using raw sockets requires running as an Administrator on Windows and the `CAP_NET_RAW` capability on Linux

**Comment by webknjaz** [comment]
> unfortunately, using `SO_LINGER` to send a `RST` in the middle of a `FIN` shutdown is nearly impossible as it becomes a timing issue

yeah.. that's what I've concluded after some experiments.


> note that using raw sockets requires running as an Administrator on Windows and the `CAP_NET_RAW` capability on Linux

Woah, this adds a lot of complications. I'll have to postpone this idea and just simulate the exceptions with a `Mock(side_effect=)` where possible (https://github.com/cherrypy/cheroot/compare/testing/341-conn-reset-on-sock-shutdown-regression).

P.S. Thanks for the feedback! I did have in mind some ideas that I wanted to emit RST in a controlled way but never thought about the raw sockets... I also wanted to see if [mitmproxy](https://docs.mitmproxy.org/stable/concepts-protocols/#raw-tcp--tcp-proxy--fallback) would help me achieve the required result but it doesn't look like it'd be helpful.

**Pull request merged**